Update Operator README for release (#370)
minio/operator: README.md

# MinIO Operator [![Docker Pulls](https://img.shields.io/docker/pulls/minio/k8s-operator.svg?maxAge=604800)](https://hub.docker.com/r/minio/k8s-operator)

<a href="https://min.io"> <img src=https://raw.githubusercontent.com/minio/minio/master/.github/logo.svg width="600px"/> </a>

> MinIO provides Kubernetes-native high performance object storage for private and

Author: Will need @abperiasamy to comment on whether this wording makes sense.

public cloud infrastructure (Hybrid Cloud). The MinIO API maintains compatibility
with the S3-API to simplify migration of applications onto MinIO Object Storage.

## Table of Contents

> * [Architecture](https://github.com/minio/operator#Architecture)

Author: Hm. These links are broken. Might need to make them explicit as `operator/tree/master/README.md#` and see if that works...

* [Create a MinIO Tenant](https://github.com/minio/operator#Create-a-MinIO-Tenant)
* [Expand a MinIO Tenant](https://github.com/minio/operator#Expand-a-MinIO-Tenant)
* [Kubernetes Cluster Configuration](https://github.com/minio/operator#Kubernetes-Cluster-Configuration)

# Architecture

Each MinIO Tenant represents an independent MinIO Object Store service within
the Kubernetes cluster. The following diagram describes the architecture of as
MinIO Tenant deployed into Kubernetes:

> IMAGE

Author: Will create diagram if we first decide this looks OK.

MinIO provides multiple methods for accessing and managing the MinIO Tenant:

## MinIO Console

The MinIO Console provides a graphical user interface (GUI) for interacting with
MinIO Tenants. 

> IMAGE

Author: ToDo: Update section for when Console for Operator (Tenant Management) is ready to include something like:

"Kubernetes Administrators can create and manage MinIO Tenants entirely through the Console and without relying on command line interactions." or something.

## MinIO Operator and `kubectl` Plugin

The MinIO Operator extends the Kubernetes API to support deploying MinIO-specific
resources as a Tenant in a Kubernetes cluster.

The MinIO `kubectl minio` plugin wraps the Operator to provide a simplified interface
for deploying and managing MinIO Tenants in a Kubernetes cluster through the
`kubectl` command line tool.

# Create a MinIO Tenant

This procedure creates a 4-node MinIO Tenant suitable for evaluation and 
early development using MinIO for object storage.

- MinIO requires Kubernetes version 1.17.0 or later.

- This procedure assumes the cluster contains a 
  [namespace](https://github.com/minio/operator#create-a-namespace) for
  the MinIO Tenant.

- This procedure assumes the cluster contains a
  [`StorageClass`](https://github.com/minio/operator#default-storage-class)
  for the MinIO Tenant Persistent Volume Claims (`PVC`).

## 1) Install the MinIO Operator

### Install using `kubectl krew`

Run the following command to install the MinIO Operator and Plugin using `krew`:

```sh
   kubectl krew update
   kubectl krew install minio
```

Run the following command to initialize the Operator:

```sh
kubectl minio init

> 

Reviewer: This is actually `kubectl minio operator init`, can you fix this as well pls @ravindk89

  deployed as part of the Tenant.

- The `--servers` field indicates the number of `minio` pods to deploy into the cluster.
  The cluster *must* have at least one available Node per `minio` pod. 

- The `--volumes` field indicates the total number of volumes in the Tenant. MinIO
  generates a Persistent Volume Claim (`PVC`) for each volume and evenly distributes
  volumes across each `minio` pod. The example above results in 4 volumes per `minio` pod.

- The `--capacity` field indicates the total capacity of the cluster. MinIO determines the
  amount of storage to request for each `pvc` by dividing the specified capacity by the
  total number of volumes in the server. The example above results in 1Ti requested
  capacity per volume.

- The `--namespace` field indicates the namespace onto which MinIO deploys the Tenant. 
  If omitted, MinIO uses the `Default` namespace.

## 3) Connect to the Tenant

MinIO outputs credentials for connecting to the MinIO Tenant as part of the creation
process:

```sh
MinIO Tenant 'minio-tenant1' created

Username: admin 
Password: abcd1234-abcd-1234-abcd-1234abcd1234

Web interface access: 

> 	$ kubectl port-forward svc/minio-tenant1-console 9090:9443
> 	Point browser to https://localhost:9090

Reviewer 2: ```suggestion
	$ kubectl port-forward svc/minio-tenant1-console 9443:9443
	Point browser to https://localhost:9443
```

Here we can use the same `9443` port

Object storage access: 

> 	$ kubectl port-forward svc/minio-tenant1-hl 9000:9000

Reviewer 2: ```suggestion
	$ kubectl port-forward svc/minio 9000:9000
```
TLDR; xd - Here service name should be only `minio`

When operator deploys a new tenant, 2 services are created, "minio" clusterIP service (has an IP address assigned) and "minio-tenant1-hl" headless service (doesnt have an IP address assigned), ie:

```
kubectl get svc
NAME                                           TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE
kubernetes                                     ClusterIP   10.96.0.1       <none>        443/TCP             19d
minio                                          ClusterIP   10.98.115.236   <none>        443/TCP             24m
minio-tenant1-console   ClusterIP   10.97.71.11     <none>        9090/TCP,9443/TCP   23m
minio-tenant1-hl        ClusterIP   None            <none>        9000/TCP            24m
```
It's better for applications to access directly the `minio` service (K8S will LB connections to all minio pods) instead of the headless service, otherwise they will need to implement LB logic in the client side https://medium.com/faun/kubernetes-headless-service-vs-clusterip-and-traffic-distribution-904b058f0dfd

Author: @nitisht @Alevsk  this might require a change to the operator PR

This was based on the output from #369 [here](https://github.com/minio/operator/pull/369/files#diff-22fba376bed03b38062f292edd3f54d976d78364ab8a1f3688f5b830acdd0a51R174), so maybe `minioSVCName` needs to be modified?

Author: This should now be addressed.

	$ mc alias set tenant1 https://localhost:9000 admin abcd1234-abcd-1234-abcd-1234abcd1234
```

The `kubectl port-forward` command temporarily forwards traffic
from the local host to the MinIO Tenant. 

- The `Web interface access` opens the MinIO Console interface for interacting
  with and managing the MinIO Tenant.

- The `Object storage access` provides both the classic MinIO web interface *and* general 
  API access to the MinIO Object Store.

Applications inside the Kubernetes cluster can access the MinIO Tenant through either the
web interface service (`minio-tenant1-console`) *or* the object storage service

> (`minio-tenant1-hl`). 

Reviewer 2: ```suggestion
(`miniol`). 
```

The following `kubectl minio` command expands a MinIO Tenant with an additional
4 nodes, 16 volumes, and added capacity of 16Ti. This configuration requires
*at least* 16 [Persistent Volumes](https://github.com/minio/operator#Local-Persistent-Volumes).

```sh

   kubectl minio tenant expand \
      --name minio-tenant1 \
      --servers 4 \
      --volumes 16 \
      --capacity 16Ti

```

- The `--name` field specifies the name of the existing MinIO Tenant to expand.

- The `--servers` field indicates the number of `minio` pods to deploy into the cluster.
  The cluster *must* have at least one available Node per `minio` pod. 

- The `--volumes` field indicates the total number of volumes in the Tenant. MinIO
  generates a Persistent Volume Claim (`PVC`) for each volume and evenly distributes
  volumes across each `minio` pod. The example above results in 4 volumes per `minio` pod.

- The `--capacity` field indicates the total capacity of the cluster. MinIO determines the
  amount of storage to request for each `pvc` by dividing the specified capacity by the
  total number of volumes in the server. The example above results in 1Ti requested
  capacity per volume.

> # Kubernetes Cluster Configuration

Author: Might be a good place to also stick in something around networking, like

## Networking Recommendations

MinIO can make efficient use of storage and can completely saturate network interfaces.

For example, consider a Node with locally-attached NVME SSD storage. The SSD can exhibit a maximum READ/WRITE performance of 2Gbps. If the node has 8 such drives, MinIO may drive a theoretical total of 16Gbps of traffic and overwhelm a 10GbE NIC. 

To avoid network bottlenecks, consider calculating the total possible READ/WRITE throughput of a single node and compare it against the total available network bandwidth provided by the node's NIC *and* the network infrastructure between that NIC and Clients. You can safely assume MinIO can drive up to 85% of theoretical drive throughput. 

Spitballing heavily here.

# MinIO Operator [![Docker Pulls](https://img.shields.io/docker/pulls/minio/k8s-operator.svg?maxAge=604800)](https://hub.docker.com/r/minio/k8s-operator)

<a href="https://min.io"> <img src=https://raw.githubusercontent.com/minio/minio/master/.github/logo.svg width="600px"/> </a>

MinIO provides Kubernetes-native high performance object storage for private and
public cloud infrastructure (Hybrid Cloud). The MinIO API maintains compatibility
with the S3-API to simplify migration of applications onto MinIO Object Storage.

## Table of Contents

* [Architecture](https://github.com/minio/operator/blob/master/README.md#Architecture)
* [Create a MinIO Tenant](https://github.com/minio/operator/blob/master/README.md#Create-a-MinIO-Tenant)
* [Expand a MinIO Tenant](https://github.com/minio/operator/blob/master/REAMDE.md#Expand-a-MinIO-Tenant)
* [Kubernetes Cluster Configuration](https://github.com/minio/operator/blob/master/README.md#Kubernetes-Cluster-Configuration)

# Architecture

> Each MinIO Tenant represents an independent MinIO Object Store service within

Author: Done

the Kubernetes cluster. The following diagram describes the architecture of as
MinIO Tenant deployed into Kubernetes:

IMAGE

MinIO provides multiple methods for accessing and managing the MinIO Tenant:

## MinIO Console

The MinIO Console provides a graphical user interface (GUI) for interacting with
MinIO Tenants. 

IMAGE

> MinIO Tenant administrators can perform a variety of tasks through the Console,

Reviewer 3: Did we define Tenant administrators somewhere?  If not this might cause confusion, can probably switch slightly to "Administrators of MinIO tenants"

Author: Done

[MinIO Console Github Repository](https://github.com/minio/console).

## MinIO Operator and `kubectl` Plugin

The MinIO Operator extends the Kubernetes API to support deploying MinIO-specific
resources as a Tenant in a Kubernetes cluster.

The MinIO `kubectl minio` plugin wraps the Operator to provide a simplified interface
for deploying and managing MinIO Tenants in a Kubernetes cluster through the
`kubectl` command line tool.

# Create a MinIO Tenant

This procedure creates a 4-node MinIO Tenant suitable for evaluation and 
early development using MinIO for object storage.

- MinIO requires Kubernetes version 1.17.0 or later.

- This procedure assumes the cluster contains a 
  [namespace](https://github.com/minio/operator/blob/master/README.md#create-a-namespace) for
  the MinIO Tenant.

- This procedure assumes the cluster contains a
  [`StorageClass`](https://github.com/minio/operator/blob/master/README.md#default-storage-class)
  for the MinIO Tenant Persistent Volumes  (`PV`). The `StorageClass`
  *must* have `volumeBindingMode: WaitForFirstConsumer`

## 1) Install the MinIO Operator

### Install using `kubectl krew`

> Run the following command to install the MinIO Operator and Plugin using `krew`:

Reviewer 3: Do we assume krew is a given within an administrators environment?  If not we should probably link to download or docs for those who are unfamiliar

Author: Added as part of the pre-requisites.

```sh
   kubectl krew update
   kubectl krew install minio
```

Run the following command to initialize the Operator:

```sh
kubectl minio init

```

## 2) Create a New Tenant

The following `kubectl minio` command creates a MinIO Tenant with 4 nodes, 16
volumes, and a total capacity  of 16Ti. This configuration requires
*at least* 16 
[Persistent Volumes](https://github.com/minio/operator#Local-Persistent-Volumes).

```sh
   kubectl minio tenant create minio-tenant-1 \
      --servers 4                             \
      --volumes 16                            \
      --capacity 16Ti                         \
      --namespace minio-tenant-1              \
      --storageClassName local-storage        \
```

- The `minio-tenant-1` argument specifies the name of the MinIO Tenant. The MinIO
  Operator uses this name as a prefix for certain resources in the Tenant.

- The `--servers` field indicates the number of `minio` pods to deploy into the cluster.

>   The cluster *must* have at least one available Node per `minio` pod. 

Reviewer 3: Do we need to be pedantic and add "worker node" here?

Author: I think it's worth being that specific. The Kubernetes docs uses the two terms interchangeably, but might as well avoid that potential edge.

- The `--volumes` field indicates the total number of volumes in the Tenant. MinIO
  generates a Persistent Volume Claim (`PVC`) for each volume and evenly distributes
  volumes across each `minio` pod. The example above results in 4 volumes per `minio` pod.

>   If the cluster does not have sufficient Persistent Volume (`PV`) to bind to the

Reviewer 3: ```suggestion
  If the cluster does not have sufficient Persistent Volumes (`PV`) to bind to the
```

>   generated `PVC`, Tenant creation hangs until the required `PV` exist.

Reviewer 3: ```suggestion
  generated `PVC`, Tenant creation hangs until the required `PV` exists.
```

Author: So I *think* that here, if we define `PV` as "Persistent Volumes", then "exist" would be correct. But its so damn ugly, that I'm more tempted to re-write this as

"If the cluster does not have an unbound Persistent Volume (`PV`) for each generated `PVC`, Tenant creation hangs until the required number of `PV`s exist."

`PV`s reads so weird but acronyms have that problem anyways.

Author: Tweaked to avoid this grammatical oddity entirely.

- The `--capacity` field indicates the total capacity of the cluster. MinIO determines the
  amount of storage to request for each `pvc` by dividing the specified capacity by the
  total number of volumes in the server. The example above results in 1Ti requested
  capacity per volume.

>   If the cluster does not have sufficient Persistent Volume (`PV`) to satsify the

Reviewer 3: ```suggestion
  If the cluster does not have sufficient Persistent Volumes (`PV`) to satisfy the
```

>   requested capacity, Tenant creation hangs until the required `PV` exist.

Reviewer 3: ```suggestion
  requested capacity, Tenant creation hangs until the required `PV` exists.
```

Author: Same as above. I'm actually not sure of the exact grammar rules here (how do you pluralize an acronym anyways). Will chew on this.

Author: Tweaked as per above

- The `--namespace` field indicates the namespace onto which MinIO deploys the Tenant. 
  If omitted, MinIO uses the `Default` namespace.

  MinIO supports *one* MinIO Tenant per namespace.

> - The `--storageClassName` field indicates which `StorageClass` to use when generating

Reviewer 3: Link to storage class docs here?

Author: Done

+-------------+------------------------+------------------+--------------+-----------------+

```

Copy the credentials to a secure location, such as a password protected key manager. 
MinIO does *not* display these credentials again.

MinIO Tenants deploy with TLS enabled by default, where the MinIO Operator uses the 
Kubernetes `certificates.k8s.io` API to generate the required x.509 certificates. Each 
certificate is signed using the Kubernetes Certificate Authority (CA) configured during
cluster deployment. While Kubernetes mounts this CA on Pods in the cluster, Pods do 
*not* trust that CA by default. You must copy the CA to a directory such that the 
`update-ca-certificates` utility can find and add it to the system trust store to 
enable validation of MinIO TLS certificates:

```sh

cp /var/run/secrets/kubernetes.io/serviceaccount/ca.crt /usr/local/share/ca-certificates/
update-ca-certificates
```

For applications *external* to the Kubernetes cluster, you must configure 
[Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) or a
[Load Balancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) to

> expose the MinIO Tenant services. The `kubectl port-forward` command temporarily forwards traffic

Reviewer: ```suggestion
expose the MinIO Tenant services. Alternatively, you can use the `kubectl port-forward` command to temporarily forward traffic 
```

Author: Resolved in separate commit.

from the local host to the MinIO Tenant. 

- The `minio` service provides access to MinIO Object Storage operations.

- The `minio-tenant-1-console` service provides access to the MinIO Console. The 

>   MinIO Console supports administration of the MinIO Tenant.

Reviewer 3: ```suggestion
  MinIO Console supports GUI administration of the MinIO Tenant.
```

# Expand a MinIO Tenant

MinIO supports expanding an existing MinIO Tenant onto additional hosts and storage.

- MinIO requires Kubernetes version 1.17.0 or later.

- This procedure assumes the cluster contains a 
  [namespace](https://github.com/minio/operator#create-a-namespace) for
  the MinIO Tenant.

- This procedure assumes the cluster contains a
  [`StorageClass`](https://github.com/minio/operator#default-storage-class)
  for the MinIO Tenant Persistent Volume Claims (`PVC`).

The following `kubectl minio` command expands a MinIO Tenant with an additional
4 nodes, 16 volumes, and added capacity of 16Ti. This configuration requires

> *at least* 16 [Persistent Volumes](https://github.com/minio/operator#Local-Persistent-Volumes).

Reviewer 3: Should this say 16 new?  Also, it is possible for readers to misconstrue the requirement of 16 as a hard limit in minio rather than just what is required for this example

Author: Made some tweaks to hopefully make it clear this is driven by the number of PVC

strongly recommends creating a custom `StorageClass` for use by 
`PV` supporting a MinIO Tenant:

The following `StorageClass` object contains the appropriate fields for use with the MinIO Plugin:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
    name: local-storage
provisioner: kubernetes.io/no-provisioner
volumeBindingMode: WaitForFirstConsumer
```

To specify the storage class, include the `--storageClassName` option to 
`kubectl minio tenant create`.

## Local Persistent Volumes

MinIO automatically creates Persistent Volume Claims (PVC) as part of Tenant creation.
Ensure the cluster at least one 
[Persistent Volume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
for each PVC MinIO requests.

You can estimate the number of PVC by multiplying the number of nodes in the 
Tenant by the number of drives per node. For example, a 4-node Tenant with
4 drives per node requires 16 PVC and therefore 16 PV.

MinIO *strongly recommends* using locally-attached storage for each PV for
object storage performance. MinIO recommends the following CSI drivers for
creating local PV:

> 

Reviewer: ```suggestion
MinIO *strongly recommends* recommends the following CSI drivers for creating local PV:
```

Author: Resolved in separate commit to crunch two lines together.

# Kubernetes Cluster Configuration

## Default Storage Class

The MinIO Kubernetes Plugin (`kubectl minio`) automatically generates
Persistent Volume Claims (`PVC`) as part of deploying a MinIO Tenant. 
The plugin defaults to creating each `PVC` with the `default`
Kubernetes [`Storage Class`](https://kubernetes.io/docs/concepts/storage/storage-classes/).

MinIO Tenants *require* that the `StorageClass` set 
`volumeBindingMode` to `WaitForFirstConsumer`. The default `StorageClass` may use the 
`Immediate` setting, which can cause complications during `PVC` binding. MinIO 
strongly recommends creating a custom `StorageClass` for use by 
`PV` supporting a MinIO Tenant:

The following `StorageClass` object contains the appropriate fields for use with the MinIO Plugin:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
    name: local-storage
provisioner: kubernetes.io/no-provisioner
volumeBindingMode: WaitForFirstConsumer
```

To specify the storage class, include the `--storageClassName` option to 
`kubectl minio tenant create`.

## Local Persistent Volumes

MinIO automatically creates Persistent Volume Claims (PVC) as part of Tenant creation.

> Ensure the cluster at least one 

Reviewer 3: ```suggestion
Ensure the cluster has at least one 
```

[Persistent Volume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
for each PVC MinIO requests.

> You can estimate the number of PVC by multiplying the number of `minio` server pods  in the 

Reviewer 3: ```suggestion
You can estimate the number of PVC by multiplying the number of `minio` server pods in the 
```
